finagle-memcached: Bugfix memcached Zipkin integration

kristofa opened the pull request:
Problem

When getting a value from memcached the cache hit/miss annotation
was submitted after the cr annotation was submitted.

Submitting the cache hit/miss annotation after the cr annotation resulted
in the span being recreated and ending up in the DeadlineSpanMap.
When the deadline exceeded the span was submitted a 2nd time overwriting
the correct duration of the original span.  Once that happened
the span was also annotated with 'finagle.flush'.

Solution

Update the memcached specific tracing filter so memcached specific
annotations are submitted in between cs and cr annotations.

Result

The memcached span is not re-submitted a 2nd time and does not end
up in DeadlineSpanMap. The duration is correct when viewed in zipkin-web
and there is no 'finagle.flush' annotation anymore.

kristofa commented:
@mosesn 2nd attempt to fixing #343.

kristofa commented:
![screen shot 2016-06-13 at 9 25 48 am](https://cloud.githubusercontent.com/assets/2221492/16000076/2c598f82-314a-11e6-9483-ea29e4b8ebd0.png)

This screenshot from zipkin-web shows the behavior of the original code. While the memcached 'get' operation didn't even take 1ms the duration of the span ends up being more than 32 seconds because the span is submitted a 2nd time (finagle.flush annotation).

I tested the fix I submit here using a test with a local memcached and zipkin set-up.

dschobel commented on finagle-memcached/src/main/scala/com/twitter/finagle/Memcached.scala:
```diff
@@ -22,32 +22,31 @@ import com.twitter.finagle.tracing._
-        val response = service(command)
+  object TracingFilter extends SimpleFilter[Command, Response] {
+    def apply(command: Command, service: Service[Command, Response]) = {
+      val response = service(command)
```
doesn't sequencing this request dispatch before the rpc trace below leave the trace out of order with regards to any tracing lower in the stack?

kristofa replied:
I think we are fine with these changes.
The order in which tracing events happen for memcached with this commit in place:

1. Initialize span (TraceInitializerFilter part of StackClient)
2. Add Client Send annotation (AnnotatingTracingFilter set-up in MemcachedTracingFilter.Module)
3. Execute memcached request
4. Add cache hit/miss annotations based on memcached response (MemcachedTracingFilter)
5. Add Client Received annotation (AnnotatingTracingFilter set-up in MemcachedTracingFilter.Module)

We can only add the cache hit/miss annotations once we got the response. The important change is that the memcached annotations are submitted in between the cs / cr annotations. The Client Received annotation should be the last annotation that gets submitted because it indicates the span is complete and ready to send off to back-end.

dschobel replied:
Step #3 represents the rest of the client stack and we have no guarantees
that there's no tracing happening there so ideally we'd do the request
level tracing before dispatching the request through the rest of the stack.

I'm okay with leaving it as-is but can you leave a comment that this
implementation assumes no further tracing lower in the stack?

On Fri, Jun 17, 2016 at 1:04 AM, Kristof Adriaenssens <
notifications@github.com> wrote:

> In finagle-memcached/src/main/scala/com/twitter/finagle/Memcached.scala
> <https://github.com/twitter/finagle/pull/516#discussion_r67473223>:
>
> >      }
> >    }
> >
> > -  class Filter(tracer: Tracer) extends SimpleFilter[Command, Response] {
> > -    def apply(command: Command, service: Service[Command, Response]): Future[Response] =
> > -      Trace.letTracerAndNextId(tracer) {
> > -        val response = service(command)
> > +  object TracingFilter extends SimpleFilter[Command, Response] {
> > +    def apply(command: Command, service: Service[Command, Response]) = {
> > +      val response = service(command)
>
> I think we are fine with these changes.
> The order in which tracing events happen for memcached with this commit in
> place:
>
>    1. Initialize span (TraceInitializerFilter part of StackClient)
>    2. Add Client Send annotation (AnnotatingTracingFilter set-up in
>    MemcachedTracingFilter.Module)
>    3. Execute memcached request
>    4. Add cache hit/miss annotations based on memcached response
>    (MemcachedTracingFilter)
>    5. Add Client Received annotation (AnnotatingTracingFilter set-up in
>    MemcachedTracingFilter.Module)
>
> We can only add the cache hit/miss annotations once we got the response.
> The important change is that the memcached annotations are submitted in
> between the cs / cr annotations. The Client Received annotation should be
> the last annotation that gets submitted because it indicates the span is
> complete and ready to send off to back-end.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/twitter/finagle/pull/516/files/67af7342d082c4408f125b8d15cf68a394506b2a#r67473223>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AADkAZf67j9RBSfbjaWWpioBOLc6H9-Dks5qMlUDgaJpZM4I0Dkv>
> .
>

kristofa replied:
I certainly want to add a comment but I would like to understand the problem. 

What could happen is that additional tracing annotations are added lower in the memcached client stack but I don't see why this would give a problem with adding the cache hit/miss annotations on the way out? 

Thank you for the quick feedback!

dschobel replied:
Sorry I should have been more specific, the hit/miss annotations are fine, the issue is sequencing `Trace.recordRpc(command.name)` in line 47 after potential tracing lower in the stack. It seems like we'd want the cache command name to be traced before the remainder for the stack which is dispatching the command.

dschobel commented on finagle-memcached/src/main/scala/com/twitter/finagle/Memcached.scala:
```diff
@@ -22,32 +22,31 @@ import com.twitter.finagle.tracing._
         command match {
-          case command: RetrievalCommand if Trace.isActivelyTracing =>
+          case command: RetrievalCommand =>
             response onSuccess {
```
no infix notation please.

dschobel commented on finagle-memcached/src/main/scala/com/twitter/finagle/Memcached.scala:
```diff
@@ -58,13 +57,18 @@ private[finagle] object MemcachedTraceInitializer {
                   misses.remove(key)
                 }
-                misses foreach { Trace.recordBinary(_, "Miss") }
+                misses foreach {
```
ditto, no infix

dschobel commented on finagle-memcached/src/main/scala/com/twitter/finagle/Memcached.scala:
```diff
@@ -58,13 +57,18 @@ private[finagle] object MemcachedTraceInitializer {
                 }
-                misses foreach { Trace.recordBinary(_, "Miss") }
+                misses foreach {
+                  Trace.recordBinary(_, "Miss")
```
pulling this trace call into a function literal shaves an allocation on the request path

dschobel commented:
thanks for the patch, one question about traces lower in the stack, otherwise just nits.

kristofa commented:
@dschobel I updated the code to submit the rpc name (command name) before dispatching the request.

dschobel commented:
We still need to guard the traceRpc call with isActivelyTracing so I think
a comment on the previous version is fine.

On Wed, Jun 22, 2016 at 1:03 PM, Kristof Adriaenssens <
notifications@github.com> wrote:

> @dschobel <https://github.com/dschobel> I updated the code to submit the
> rpc name (command name) before dispatching the request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/twitter/finagle/pull/516#issuecomment-227860137>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AADkATbGYTscu24bO2JoT_s0D50WxXV5ks5qOZUQgaJpZM4I0Dkv>
> .
>

kristofa commented:
Inside 'traceRpc' there is a verification using 'isActivelyTracing' before submitting the annotation. 
But we should have no tracing further down the stack with memcached anyway so I reverted to previous version and added a comment as you suggested.

dschobel commented:
@kristofa ah good point, I didn't see the isTracing guard in record. anyway, I'm happy with this as-is. thanks again for the patch. 

lgtm

kristofa commented:
@dschobel Thanks for reviewing! I hope this gets merged as it will make our trace overviews usable again.

dschobel commented:
@kristofa it's been posted internally for review so I expect it'll merge and get synced out on Monday.

mosesn commented:
@kristofa this was published to `develop` https://github.com/twitter/finagle/commit/5c8862984b9f01e6863459d8bba4f548981b2629, and we're working on a release now.  thanks for the contribution!

kristofa commented:
I'm happy that I'm able to contribute. Thanks for open sourcing such a great software project!
